### ISSUE TITLE
Document streaming usecase (like `UNBOUNDED` tables)

### ISSUE THREAD

**Issue by alamb** [apache/arrow-datafusion#9016]
### Is your feature request related to a problem or challenge?

Someone asked in [discord](https://discord.com/channels/885562378132000778/1166447479609376850/1200557140096594053): 

> I'm looking at Datafusion and Polars as potential solutions for calculating averages over a sliding window of events, where the window is bound by event time. I've just come across Datafusion, would anyone be able to clarify if it's suitable for this use case? In essence, I have events streaming in via RPC that I want to feed into a a system that gives the above outcome. 

I am pretty sure this is exactly the case for using `UNBOUNDED` tables with explicitly defined `ORDER BY` from synnada and others.  However, when I went to look for the documentation, I could't find any mention of this usecase or the unbounded tables

### Describe the solution you'd like

I would like to help make it easier for people to use DataFusion for streaming usecases by:
- [ ]  Documenting the `UNBOUNDED` keyword [in the `CREATE EXTERNAL TABLE` documentation](https://arrow.apache.org/datafusion/user-guide/sql/ddl.html#create-external-table)
- [ ] Add an example in https://github.com/apache/arrow-datafusion/tree/main/datafusion-examples/examples with a simple streaming example (perhaps implementing some simple version of the use case described in the description)
- [ ] Add a [section to the library guide](https://arrow.apache.org/datafusion/library-user-guide/index.html) giving some basic

**Comment by alamb** [comment]
@metesynnada @mustafasrepo  @ozankabak or @edmondop @mwylde do you know of any existing documentation / examples we could adapt?

**Comment by alamb** [comment]
I think the major initial push in this area came in  https://github.com/apache/arrow-datafusion/pull/4694

**Comment by Tangruilin** [comment]
I can help with it.

Assign to me. @alamb

**Comment by edmondop** [comment]
Would one need to write custom unbounded sources?

**Comment by alamb** [comment]
> Would one need to write custom unbounded sources?

I don't think so @edmondop  -- I was thinking anything that gives others examples / help starting would be great. Maybe we can start with some SQL reference / API docs and high level commentary in one PR and then add an example as another PR

Right now there is nothing documented, so it will be very easy to imprve the status quo!

**Comment by trungda** [comment]
I've realized that `StreamEncoding` is not supported for Parquet. Is it intentional?

**Comment by alamb** [comment]
> : Is there any trick out there for watermarking, that is, a way for a table to emit a signal that all events at a given event time have been sent?

The way most of the code in DataFusion works is that it will use the next distinct value in the data to trigger emission (as in you have to see an event of the next time)

I don't know of any way to send a synthetic signal that says "will never see any more values in this time interval"

**Comment by ozankabak** [comment]
> I don't know of any way to send a synthetic signal that says "will never see any more values in this time interval"

Yes, IMO this seems to fall out of scope for "upstream" Datafusion (even though we also have these challenges but choose to solve them downstream due to this reasoning).

**Comment by milenkovicm** [comment]
A bit late to this party :), few questions regarding this topic:

1. I would tell that `UNBOUNDED` is intended to be used with queries like `INSERT INTO kafka_table_1 SELECT * FROM kafka_table_2`, do I get this correctly? 
2. if 1 is correct, I believe we would have to detach `ctx.sql(...)` from the main thread, not to block other statements, so the question is who will be in charge of query cancelation?  Would it be/ is it part of datafusion, or it has to be implemented as part of the source, or somewhere else 

thanks a lot

**Comment by alamb** [comment]
> 2\. if 1 is correct, I believe we would have to detach `ctx.sql(...)` from the main thread, not to block other statements, so the question is who will be in charge of query cancelation?  Would it be/ is it part of datafusion, or it has to be implemented as part of the source, or somewhere else

I think the execution is run as a separate tokio task -- so whenever someone wants to see if the next result is available they would do `stream.next().await` or something equivalent

**Comment by milenkovicm** [comment]
wrt point ,. if we have a running unbounded pipeline, when shutdown/cancel event comes, sources should stop producing, probably commit offsets and so on and give chance for rest of the pipeline to finish all the bits in-flight before shutdown. 

so if i can reformulate my question 2, "who will inform source that it should wrap up for this execution and shut down ? :)"

or am I missing something obvious here?

**Comment by alamb** [comment]
> so if i can reformulate my question 2, "who will inform source that it should wrap up for this execution and shut down ? :)"

I think this needs to be handled by the containing system (nothing in DataFusion will do it directly). It would have to hold on to some reference to the source that can be signaled to stop.

In terms of shutdown,  when the stream is `drop`ed, it will shutdown all resources attached to the executing plan. This is described a bit here https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#cancellation--aborting-execution. However, this doesn't do anything to save state as would be required in a streaming system

**Comment by alamb** [comment]
Some background on the current state of how a data source declares itself unbounded / streaming, which such documentation could cover.

There are three layers, from lowest to highest:

**1. The fundamental mechanism — `PlanProperties` on any `ExecutionPlan`**

A source (or any operator) declares itself unbounded by returning `Boundedness::Unbounded { requires_infinite_memory: false }` in the [`PlanProperties`](https://docs.rs/datafusion/latest/datafusion/physical_plan/struct.PlanProperties.html) it builds in `properties()` (the constructor takes `boundedness`; there is also `with_boundedness()`):

- [`Boundedness`](https://docs.rs/datafusion/latest/datafusion/physical_plan/execution_plan/enum.Boundedness.html) and [`EmissionType`](https://docs.rs/datafusion/latest/datafusion/physical_plan/execution_plan/enum.EmissionType.html), added in #13823
- Accessors: [`ExecutionPlanProperties::boundedness()`](https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlanProperties.html#tymethod.boundedness) and [`pipeline_behavior()`](https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlanProperties.html#tymethod.pipeline_behavior)

**2. The ready-made physical operator — `StreamingTableExec`**

The intended way to expose a custom stream: implement `PartitionStream` and pass `infinite: true` to [`StreamingTableExec::try_new`](https://docs.rs/datafusion/latest/datafusion/physical_plan/streaming/struct.StreamingTableExec.html). That flag is the *only* thing that becomes `Unbounded` — see [`compute_properties`](https://github.com/apache/datafusion/blob/128ef37d6256c278418f1425da28624816a8e95b/datafusion/physical-plan/src/streaming.rs#L171-L187) (emission is hard-coded `Incremental`).

For generated data there is also [`LazyBatchGenerator::boundedness()`](https://docs.rs/datafusion/latest/datafusion/physical_plan/memory/trait.LazyBatchGenerator.html) (default `Bounded`, override to declare unbounded).

**3. The `TableProvider` / SQL layer**

- [`StreamingTable`](https://docs.rs/datafusion/latest/datafusion/catalog/streaming/struct.StreamingTable.html) with `.with_infinite_table(true)` (default is `false`)
- In SQL: `CREATE UNBOUNDED EXTERNAL TABLE` routes to `StreamTableFactory`, whose [`scan()` hardcodes `infinite = true`](https://github.com/apache/datafusion/blob/128ef37d6256c278418f1425da28624816a8e95b/datafusion/catalog/src/stream.rs#L344-L351) — user guide: https://datafusion.apache.org/user-guide/sql/ddl.html

Two gaps worth noting:

- `DataSourceExec` (i.e., every `FileScanConfig`-based file source) *cannot* declare unboundedness — it [hardcodes `Boundedness::Bounded`](https://github.com/apache/datafusion/blob/128ef37d6256c278418f1425da28624816a8e95b/datafusion/datasource/src/source.rs#L597-L605)
- [`JsonReadOptions::infinite` / `mark_infinite()`](https://github.com/apache/datafusion/blob/128ef37d6256c278418f1425da28624816a8e95b/datafusion/core/src/datasource/file_format/options.rs#L475) appears to be dead code with no consumer

There is currently no narrative "how to build a streaming source" doc — the closest existing docs are the [`StreamingTableExec` rustdoc](https://docs.rs/datafusion/latest/datafusion/physical_plan/streaming/struct.StreamingTableExec.html) and the [DDL page](https://datafusion.apache.org/user-guide/sql/ddl.html). Adding this is tracked in 
- #9016

**Comment by goutamadwant** [comment]
@alamb can i give this a shot and try to complete the remaining Library User Guide documentation for streaming and unbounded execution. 
will plan to document the PlanProperties/Boundedness layer, StreamingTableExec with PartitionStream, the TableProvider/SQL entry points, cancellation behavior, and current limitations around watermarks, checkpointing and state persistence, while linking the existing streaming example and DDL documentation. would this be a good start? Let me know if you have any pointers. thanks !

**Comment by alamb** [comment]
Thanks @goutamadwant  that will be amazing. -- I think the key here will be to break the work down into manageable pieces  (so that we can review it incrementally)

I also suspect that in writing this you will uncover areas in need of additional features / cleanu, so we will have to be careful not to try and boil the ocean

What I would suggest doing first, is creating an example in datafusion-examples of the simplest possible streaming system -- basically a program that shows how to feed rows into a streaming query (perhaps something like a filter) and shows how rows come out of it as you push rows in. 

As I mentioned, I suspect this won't work great at first and well need to update some things (like the FilterExec) but starting with the example will help motivate that

With that example, we cna then start writing a "Using datafusion for streaming" section in the library user guide https://datafusion.apache.org/library-user-guide/index.html

**Comment by goutamadwant** [comment]
@alamb while prototyping the minimal streaming example, I reproduced the `FilterExec` buffering issue and opened #24353 with the focused prerequisite fix. Let me know if this makes sense. 

also to keep the work incremental, I plan to follow up with:

- A `datafusion-examples` PR demonstrating how to push batches into a streaming query and consume results incrementally.
- A library user-guide PR adding a “Using DataFusion for streaming” section based on that example.

**Comment by alamb** [comment]
Thank you @goutamadwant  -- I am sorry I haven't had a chnce to review this yet -- will do soon (hopefully tomorrow)
